core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooLock.java

```diff
@@ -123,189 +132,326 @@ public void unableToMonitorLockNode(Exception e) {
+          rValid = false;
+        }
+        // Make any invalid's sort last (higher)
+        if (lValid && rValid) {
+          int leftIdx = o1.lastIndexOf('#');
+          int rightIdx = o2.lastIndexOf('#');
+          return o1.substring(leftIdx).compareTo(o2.substring(rightIdx));
+        } else if (!lValid && rValid) {
+          return 1;
+        } else if (lValid && !rValid) {
+          return -1;
+        } else {
+          return 0;
+        }
+      }
+    });
+    if (LOG.isTraceEnabled()) {
+      LOG.trace("Children nodes");
+      for (String child : children) {
+        LOG.trace("- {}", child);
+      }
     }
+  }
 
-    List<String> children = zooKeeper.getChildren(path);
+  private synchronized void determineLockOwnership(final String myLock,
+      final AccumuloLockWatcher lw) throws KeeperException, InterruptedException {
 
-    if (!children.contains(myLock)) {
-      throw new RuntimeException("Lock attempt ephemeral node no longer exist " + myLock);
+    if (createdNodeName == null) {
+      throw new IllegalStateException(
+          "Called determineLockOwnership() when ephemeralNodeName == null");
     }
 
-    Collections.sort(children);
-    if (log.isTraceEnabled()) {
-      log.trace("Candidate lock nodes");
-      for (String child : children) {
-        log.trace("- {}", child);
+    List<String> children = new ArrayList<>();
+    zooKeeper.getChildren(path).forEach(s -> {
+      if (s.startsWith(ZLOCK_PREFIX)) {
+        children.add(s);
       }
+    });
+
+    if (!children.contains(myLock)) {
+      throw new RuntimeException("Lock attempt ephemeral node no longer exist " + myLock);
     }
+    sortChildrenByLockPrefix(children);
 
     if (children.get(0).equals(myLock)) {
-      log.trace("First candidate is my lock, acquiring");
+      LOG.trace("[{}] First candidate is my lock, acquiring", this.getZLockPrefix());
       if (!watchingParent) {
         throw new IllegalStateException(
             "Can not acquire lock, no longer watching parent : " + path);
       }
       this.lockWatcher = lw;
-      this.lock = myLock;
-      asyncLock = null;
+      this.lockNodeName = myLock;
+      createdNodeName = null;
       lockWasAcquired = true;
       lw.acquiredLock();
-      return;
-    }
-    String prev = null;
-    for (String child : children) {
-      if (child.equals(myLock)) {
-        break;
+    } else {
+      int idx = children.indexOf(myLock);
+      // Get the prefix from the prior ephemeral node
+      String prev = children.get(idx - 1);
+      int prefixIdx = prev.lastIndexOf('#');
+      String prevPrefix = prev.substring(0, prefixIdx);
+
+      // Find the lowest sequential ephemeral node with prevPrefix
+      int i = 2;
+      String lowestPrevNode = prev;
+      while ((idx - i) >= 0) {
+        prev = children.get(idx - i);
+        i++;
+        if (prev.startsWith(prevPrefix)) {
+          lowestPrevNode = prev;
+        } else {
+          break;
+        }
       }
```

Comment:
I like it how this works backwards to find the lowest prior node w/ the same uuid , its seems super efficient.  It took me a while to review the overall code for correctness and I didn't see an problems.   Could pull this into a function that could be unit tested, maybe something like `prev = findPriorLockNode(myLock,children)`.  I suggested the unit test because of the amount of time it took me to review the code.

Reply:
I can move that out into it's own function. FWIW, I'm splitting my time between two projects. I wont be working on this again until the beginning of next week so you and @ctubbsii have several days to review and leave comments for me to address when I return to this. Also, I'm working on another [test](https://github.com/dlmarion/accumulo/blob/de2e5f2377f9f5679cec0581c689bb5e14d16fea/test/src/main/java/org/apache/accumulo/test/fate/zookeeper/ZooLockIT.java#L550) that tries to lock in parallel threads.

Reply:
@keith-turner  @ctubbsii  - I have made the requested changes and have added my parallel test. I have no other changes in the works.